### PR TITLE
[BEAM-1215] Added an in memory Graphite server that allows asserting over metrics.

### PR DESCRIPTION
**Pull request by staslev**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @amitsela

### CODE REVIEW

**Comment by jbonofre** [comment]
R: @jbonofre

**Comment by staslev** [comment]
Thanks @jbonofre , let me know how it looks.

**Comment by amitsela** [comment]
It's best to have as much tests as possible, but I'm not sure the Spark runner needs to test `metrics` backend integration. I would be happy to see this as part of the SDK (maybe as a followup work to the new Metric API).
I'm afraid this will make the runner too "tight" and this also opens the question of "what about CSV sink integration ?", and more sinks contributors might add.
@staslev @jbonofre WDYT  ?

**Comment by staslev** [comment]
I think the point you've raised is indeed a concern. 
I would love to see if we can make use of this Graphite server-wannabe in integration tests in a wider scope, hopefully the recent move to the new metrics API will facilitate this.

The CSV sink is of a slightly different nature in the sense that it does not have a server backend like Graphite, so asserting over metrics sent from multiple (in-memory) executors to a CSV file might not be that useful.

Basically this leaves us with the question of whether we want to get this in now, or stash it till things get clearer in the metrics department. @amitsela @jbonofre ?

**Comment by amitsela** [comment]
@staslev care to open a discussion in dev-list ? I'm sure (at least) Ben will have something to say about this.
The new Metric API is aiming to be Dropwizard-like, so metrics-sinks (as part of the SDK) are a valid discussion IMO.

**Comment by staslev** [comment]
Sounds good, I'll shoot an email.

**Comment by davorbonaci** [comment]
Any ideas how to move this forward?

**Comment by staslev** [comment]
I think the way we view and address metrics is something that has been shaping up, but there's still things to figure out. 

I'll close this for the time being, we can reopen should the need arise.
